### PR TITLE
[Group 1] Enable nullable annotations for `Microsoft.Extensions.DependencyModel`

### PR DESCRIPTION
**Pull request by maxkoshevoi**
Related to #43605, #54012

Ready for review, but not for merge

Questions (look for `!` in the PR):
- ~~`GetNormalizedCodeBasePath.GetNormalizedCodeBasePath` is not used on net5+ due to `Assembly.CodeBase` being obsolete. Is there some alternative to preserve previous behavior, or is it ok to leave it like this?
Suppressed warning for now.~~
- It is always assumed that `reader.GetString()` returns non-null value, but it can return `null`.
Should be resolved in: https://github.com/dotnet/runtime/issues/58139
- It is always assumed that `DependencyContextJsonReader.Pool(str)` returns non-null value, but it only does that for non-null when argument. Suppressed errors with `!` for now.
Should be resolved in: https://github.com/dotnet/runtime/issues/58139
- In `AppBaseCompilationAssemblyResolver.TryResolveAssemblyPaths` `Path.GetDirectoryName(sharedPath)` might return `null`, but this is not handled.
Added `Debug.Assert`.

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs [hunk 1]
```diff
@@ -683,18 +680,18 @@ private static List<RuntimeFallbacks> ReadRuntimes(ref Utf8JsonReader reader)
...
+        private IEnumerable<Library> CreateLibraries(IEnumerable<TargetLibrary>? libraries, bool runtime, Dictionary<string, LibraryStub>? libraryStubs)
         {
             if (libraries == null)
             {
                 return Enumerable.Empty<Library>();
             }
             return libraries
                 .Select(property => CreateLibrary(property, runtime, libraryStubs))
-                .Where(library => library != null);
+                .Where(library => library != null)!;
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs:691]: Why is this necessary?

File: src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs [hunk 2]
```diff
@@ -656,7 +653,7 @@ private LibraryStub ReadOneLibrary(ref Utf8JsonReader reader)
             return new LibraryStub()
             {
                 Hash = hash,
-                Type = Pool(type),
+                Type = Pool(type)!,
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs:656]: Why is this valid?

File: src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs [hunk 3]
```diff
@@ -209,16 +210,16 @@ private DependencyContext Read(Utf8JsonReader reader)
             }
 
             return new DependencyContext(
-                new TargetInfo(framework, runtime, runtimeSignature, isPortable),
+                new TargetInfo(framework, runtime, runtimeSignature ?? string.Empty, isPortable),
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs:213]: Why this change? I believe `TargetInfo.RuntimeSignature` can be `null`.

File: src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextLoader.cs [hunk 4]
```diff
@@ -141,28 +141,30 @@ private string GetDepsJsonPath(Assembly assembly)
 
             string depsJsonFile = Path.ChangeExtension(assemblyLocation, DepsJsonExtension);
             bool depsJsonFileExists = _fileSystem.File.Exists(depsJsonFile);
-
+#if !NET5_0_OR_GREATER
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextLoader.cs:144]: Why are we making this change?

Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextLoader.cs:144]: `Assembly.CodeBase` is deprecated in .net5+: https://github.com/dotnet/runtime/issues/31127

File: src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs (hunk 1 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs:691]: Unfortunately yes. Compiler doesn't recognize `.Where(library => library != null)` as a null check.

File: src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs (hunk 3 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs:213]: It's assigned empty string in tests, so I thought, it's not meant to be `null`.

image
image

File: src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs (hunk 2 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs]: Not sure, but it looks like it is expected not to be `null`. Added `Debug.Assert` to verify.

File: src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs (hunk 1 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs:696]: Created an issue: https://github.com/dotnet/roslyn/issues/55667

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs:696]: I get it now - `CreateLibrary` can return `Library?`, so after the `.Select`, the variable is now `IEnumerable<Library?>`, and then the `Where` doesn't change the enumerable variable's type - it takes in `IEnumerable<Library?>` and returns `IEnumerable<Library?>`.

Thanks for the explanation.

File: src/libraries/Microsoft.Extensions.DependencyModel/ref/Microsoft.Extensions.DependencyModel.cs [hunk 5]
```diff
@@ -4,6 +4,8 @@
 // Changes to this file must follow the https://aka.ms/api-review process.
 // ------------------------------------------------------------------------------
 
+using System.Diagnostics.CodeAnalysis;
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyModel/ref/Microsoft.Extensions.DependencyModel.cs:7]: You shouldn't edit this file by hand. See [https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/updating-ref-source.md](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/updating-ref-source.md) for how to update it automatically.

File: src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs (hunk 3 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs:213]: I don't think we should be changing behavior here. Please keep the library's behavior the same.

File: src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs [hunk 6]
```diff
@@ -357,7 +358,7 @@ private List<Target> ReadTargets(ref Utf8JsonReader reader)
 
             while (reader.Read() && reader.IsTokenTypeProperty())
             {
-                targets.Add(ReadTarget(ref reader, reader.GetString()));
+                targets.Add(ReadTarget(ref reader, reader.GetString()!));
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs:361]: I think this `!` is lying that this can't be `null`. We should make the `ReadTarget` method accept `string? targetName`.

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs:361]: Similar comment for `ReadTargetLibrary`

File: src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextWriter.cs [hunk 7]
```diff
@@ -171,9 +171,9 @@ private void WritePortableTarget(string key, IReadOnlyList<RuntimeLibrary> runti
                     Debug.Assert(compilationLibrary.RuntimeStoreManifestName == null);
                 }
 
-                Library library = (Library)compilationLibrary ?? (Library)runtimeLibrary;
+                Library? library = (Library?)compilationLibrary ?? runtimeLibrary;
 
-                WritePortableTargetLibrary(library.Name + DependencyContextStrings.VersionSeparator + library.Version,
+                WritePortableTargetLibrary(library?.Name + DependencyContextStrings.VersionSeparator + library?.Version,
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextWriter.cs:176]: Previously this would throw a NullReferenceException if `library` was `null`. Now this will append garbage into the string.

We shouldn't change behavior here.

File: src/libraries/Microsoft.Extensions.DependencyModel/src/Utf8JsonReaderExtensions.cs [hunk 8]
```diff
@@ -14,19 +15,19 @@ internal static class Utf8JsonReaderExtensions
...
             {
-                name = reader.GetString();
+                name = reader.GetString()!;
 
                 if (reader.Read())
                 {
                     if (reader.TokenType == JsonTokenType.String)
                     {
-                        value = reader.GetString();
+                        value = reader.GetString()!;
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyModel/src/Utf8JsonReaderExtensions.cs:30]: Why is this `!` necessary?

File: src/libraries/Microsoft.Extensions.DependencyModel/src/Utf8JsonReaderExtensions.cs [hunk 9]
```diff
@@ -74,7 +75,7 @@ public static string[] ReadStringArray(this ref Utf8JsonReader reader)
 
             while (reader.Read() && reader.TokenType == JsonTokenType.String)
             {
-                items.Add(reader.GetString());
+                items.Add(reader.GetString()!);
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyModel/src/Utf8JsonReaderExtensions.cs:78]: Why is this guaranteed not to be null? I think we are lying about it here.

File: src/libraries/Microsoft.Extensions.DependencyModel/src/Utf8JsonReaderExtensions.cs (hunk 8 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.DependencyModel/src/Utf8JsonReaderExtensions.cs:30]: It's not, but as I said in PRs description, it looks like `reader.GetString()` is always assumed to be non-null.
I'll update annotations to reflect that it can return null.

File: src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs (hunk 6 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs]: Experimented with it and looks like `targetName` shouldn't be null, or code will NRE elsewhere:

- https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs#L195 (`IsRuntimeTarget` would crash if `Name` is null)
- https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs#L202 (`new TargetInfo(framework...)` a bit later would crash if `framework` is null)
- https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs#L238  (`IsRuntimeTarget` would crash if `Name` is null)

File: src/libraries/Microsoft.Extensions.DependencyModel/ref/Microsoft.Extensions.DependencyModel.cs (hunk 5 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.DependencyModel/ref/Microsoft.Extensions.DependencyModel.cs]: @eerhardt Thanks for the tip! 
`dotnet msbuild /t:GenerateReferenceAssemblySource` didn't update nullable annotations though (only attributes). Am I missing something?

File: src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs (hunk 6 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs]: I've added `Debug.Assert` here just in case

File: src/libraries/Microsoft.Extensions.DependencyModel/ref/Microsoft.Extensions.DependencyModel.cs (hunk 5 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.DependencyModel/ref/Microsoft.Extensions.DependencyModel.cs]: Once you have the nullable annotations in place in the src project, `dotnet msbuild /t:GenerateReferenceAssemblySource` from the `src` directory should update the annotations.

@krwq @joperezr @Anipik - do either of you know why they wouldn't be updated in the `ref`?

File: src/libraries/Microsoft.Extensions.DependencyModel/ref/Microsoft.Extensions.DependencyModel.cs [hunk 10]
```diff
@@ -23,26 +23,26 @@ namespace Microsoft.Extensions.DependencyModel
 {
     public partial class CompilationLibrary : Microsoft.Extensions.DependencyModel.Library
     {
-        public CompilationLibrary(string type, string name, string version, string hash, System.Collections.Generic.IEnumerable<string> assemblies, System.Collections.Generic.IEnumerable<Microsoft.Extensions.DependencyModel.Dependency> dependencies, bool serviceable) : base (default(string), default(string), default(string), default(string), default(System.Collections.Generic.IEnumerable<Microsoft.Extensions.DependencyModel.Dependency>), default(bool)) { }
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyModel/ref/Microsoft.Extensions.DependencyModel.cs:26]: Are we also going to annotate the `HashCodeCombiner` APIs above?

File: src/libraries/Microsoft.Extensions.DependencyModel/src/RuntimeAssetGroup.cs [hunk 11]
```diff
@@ -8,8 +8,8 @@ namespace Microsoft.Extensions.DependencyModel
 {
     public class RuntimeAssetGroup
     {
-        private IReadOnlyList<string> _assetPaths;
-        private IReadOnlyList<RuntimeFile> _runtimeFiles;
+        private IReadOnlyList<string>? _assetPaths;
+        private IReadOnlyList<RuntimeFile>? _runtimeFiles;
 
         public RuntimeAssetGroup(string runtime, params string[] assetPaths) : this(runtime, (IEnumerable<string>)assetPaths) { }
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyModel/src/RuntimeAssetGroup.cs:14]: It looks like `string runtime` should be `string? runtime`. Nothing is checking it for `null`.

File: src/libraries/Microsoft.Extensions.DependencyModel/ref/Microsoft.Extensions.DependencyModel.cs [hunk 12]
```diff
@@ -148,34 +148,34 @@ public partial class RuntimeAssetGroup
...
+        public TargetInfo(string framework, string? runtime, string? runtimeSignature, bool isPortable) { }
         public string Framework { get { throw null; } }
         public bool IsPortable { get { throw null; } }
-        public string Runtime { get { throw null; } }
-        public string RuntimeSignature { get { throw null; } }
+        public string? Runtime { get { throw null; } }
+        public string? RuntimeSignature { get { throw null; } }
     }
 }
 namespace Microsoft.Extensions.DependencyModel.Resolution
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyModel/ref/Microsoft.Extensions.DependencyModel.cs:181]: It looks like below 

`public AppBaseCompilationAssemblyResolver(string basePath) { }` 

should be 

`public AppBaseCompilationAssemblyResolver(string? basePath) { }`

As nothing is checking that string for null.

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.DependencyModel/ref/Microsoft.Extensions.DependencyModel.cs:181]: `AppBaseCompilationAssemblyResolver.TryResolveAssemblyPaths` could use some work too:

1. Nothing is checking `CompilationLibrary library` for `null`, but it is dereferenced. So we could start throwing `ArgNullException` when it is null.
2. `List<string> assemblies` should be `List<string>? assemblies`, since the code accepts `null`.
    - But looking at the other `ICompilationAssemblyResolver`, they don't appear to allow `null`. So maybe they all should be updated to check `assemblies` for null, and throw `ArgNull`.

Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyModel/ref/Microsoft.Extensions.DependencyModel.cs:181]: `ReferenceAssemblyPathResolver`'s constructor accepts null `string defaultReferenceAssembliesPath`.

File: src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextWriter.cs [hunk 13]
```diff
@@ -171,7 +171,7 @@ private void WritePortableTarget(string key, IReadOnlyList<RuntimeLibrary> runti
                     Debug.Assert(compilationLibrary.RuntimeStoreManifestName == null);
                 }
 
-                Library library = (Library)compilationLibrary ?? (Library)runtimeLibrary;
+                Library library = (Library?)compilationLibrary ?? runtimeLibrary ?? throw new NullReferenceException(nameof(library));
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextWriter.cs:174]: We shouldn't ever throw `NullReferenceException` explicitly. Just add a `!` here for this PR.

File: src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextLoader.cs (hunk 4 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextLoader.cs:144]: I don't think removing this code is a good idea for now. You should just suppress the Obsolete warning instead. Ex:

```C#
#pragma warning disable SYSLIB0012 // CodeBase is obsolete
```

File: src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs [hunk 14]
```diff
@@ -357,7 +358,10 @@ private List<Target> ReadTargets(ref Utf8JsonReader reader)
 
             while (reader.Read() && reader.IsTokenTypeProperty())
             {
-                targets.Add(ReadTarget(ref reader, reader.GetString()));
+                string? targetName = reader.GetString();
+                Debug.Assert(targetName != null);
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.DependencyModel/src/DependencyContextJsonReader.cs:362]: I don't think these `Debug.Assert`s are valid either. This data is coming from user input, there is no guarantee that these strings are not `null`.

Maybe the best approach here is to:

1. For this PR, just add the `!` to suppress the compiler warning.
2. Log a follow up issue to actually handle `null` being returned here. In a separate PR we can decide what to do when these strings are `null` (probably throw a decent exception), along with adding tests.

File: src/libraries/Microsoft.Extensions.DependencyModel/src/Microsoft.Extensions.DependencyModel.csproj [hunk 15]
```diff
@@ -25,11 +31,17 @@ Microsoft.Extensions.DependencyModel.DependencyContext</PackageDescription>
...
 
   <ItemGroup>
-    <PackageReference Include="System.Buffers" Version="$(SystemBuffersVersion)" />
-    <PackageReference Include="System.Memory" Version="$(SystemMemoryVersion)" />
     <ProjectReference Include="$(LibrariesProjectRoot)System.Runtime.CompilerServices.Unsafe\src\System.Runtime.CompilerServices.Unsafe.ilproj" />
     <ProjectReference Include="$(LibrariesProjectRoot)System.Text.Encodings.Web\src\System.Text.Encodings.Web.csproj" />
     <ProjectReference Include="$(LibrariesProjectRoot)System.Text.Json\src\System.Text.Json.csproj" />
   </ItemGroup>
 
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">
```
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/libraries/Microsoft.Extensions.DependencyModel/src/Microsoft.Extensions.DependencyModel.csproj:39]: > @ViktorHofer - do you know - should we only use this in the src project, and not in the ref project? Looking at the existing Extensions.* libraries, we only use it in the src projects, but use ProjectReference in the ref projects.

Yes for now please only use that setting in source projects. For reference assemblies we want P2Ps to continue to exist - at least until reference projects are deleted for OOBs (which will happen very soon).

More important, if a configuration is added to a source project, please make sure that it is also added to the reference project. That makes makes sure that there is no assembly universe mismatch and it will help with the ref<-->src convergance. It also avoids going through the netstandard shim.

**Review by eerhardt — APPROVED** [review]
Thank you again, @maxkoshevoi, for this work. This really helps!

I pushed a couple commits to this branch to get the build and unit tests passing. I also fixed a few nits and inconsistencies while I was working here.

I'll merge this once CI is green.
